File: adapters/unity/XaceInputCollector.cs
// XaceInputCollector.cs
// Collects Unity input and sends Phase 15 runtime input_packet messages.
//
// The runtime bridge consumes the schema defined in
// packages/runtime-core/src/engine_protocol.rs:
//   { msg_type, peer_id, tick, player_id, sequence_id, actions, timestamp_ms,
//     device_id, predicted }
//
// Input is sampled every Unity frame and flushed once per XACE tick. The tick
// stamp is derived from the latest runtime snapshot plus one, preserving the
// "no untimed input" replay invariant.

using System;
using System.Collections.Generic;
using UnityEngine;

namespace Xace.Adapter.Unity
{
    [RequireComponent(typeof(XaceTransport))]
    public sealed class XaceInputCollector : MonoBehaviour
    {
        [Header("Identity")]
        [SerializeField] private ulong peerId = 1;
        [SerializeField] private ulong playerId = 1;
        [SerializeField] private string deviceId = "unity";

        [Header("Ticking")]
        [SerializeField] private float fallbackTickRateHz = 60f;
        [SerializeField] private bool predicted = true;
        [SerializeField] private bool sendIdleMovement = false;
        [SerializeField] private float deadzone = 0.01f;

        [Header("Input Manager Names")]
        [SerializeField] private string horizontalAxis = "Horizontal";
        [SerializeField] private string verticalAxis = "Vertical";
        [SerializeField] private string mouseXAxis = "Mouse X";
        [SerializeField] private string mouseYAxis = "Mouse Y";
        [SerializeField] private string jumpButton = "Jump";
        [SerializeField] private string primaryFireButton = "Fire1";
        [SerializeField] private string secondaryFireButton = "Fire2";
        [SerializeField] private string interactButton = "Interact";
        [SerializeField] private string sprintButton = "Sprint";
        [SerializeField] private string crouchButton = "Crouch";

        public event Action<XaceInputPacket> OnInputPacketBuilt;

        private XaceTransport transport;
        private readonly InputAccumulator accumulator = new InputAccumulator();
        private ulong latestRuntimeTick;
        private ulong lastSentTick;
        private float tickInterval;
        private float tickAccumulator;
        private int framesAccumulated;

        private void Awake()
        {
            EnsureTransport();
            tickInterval = 1f / Mathf.Max(1f, fallbackTickRateHz);
        }

        private void OnEnable()
        {
            if (EnsureTransport())
            {
                transport.OnTickSnapshot += OnTickSnapshot;
                transport.OnHandshakeAccepted += OnHandshakeAccepted;
            }
        }

        private void OnDisable()
        {
            if (EnsureTransport())
            {
                transport.OnTickSnapshot -= OnTickSnapshot;
                transport.OnHandshakeAccepted -= OnHandshakeAccepted;
            }
        }

        private void Update()
        {
            if (!EnsureTransport() || !transport.IsConnected)
                return;

            SampleFrame();
            framesAccumulated++;

            tickAccumulator += Time.unscaledDeltaTime;
            while (tickAccumulator >= tickInterval)
            {
                tickAccumulator -= tickInterval;
                FlushTick();
            }
        }

        public void FlushNow()
        {
            if (!EnsureTransport())
                return;
            FlushTick();
        }

        private void OnHandshakeAccepted(XaceHandshakeAck ack)
        {
            if (ack.tick_rate > 0)
                tickInterval = 1f / ack.tick_rate;
        }

        private void OnTickSnapshot(XaceTickSnapshot snapshot)
        {
            latestRuntimeTick = Math.Max(latestRuntimeTick, snapshot.tick);
        }

        private void SampleFrame()
        {
            accumulator.moveX += SafeAxis(horizontalAxis);
            accumulator.moveY += SafeAxis(verticalAxis);
            accumulator.lookX += SafeAxis(mouseXAxis);
            accumulator.lookY += SafeAxis(mouseYAxis);

            accumulator.jump |= SafeButton(jumpButton);
            accumulator.primaryFire |= SafeButton(primaryFireButton);
            accumulator.secondaryFire |= SafeButton(secondaryFireButton);
            accumulator.interact |= SafeButtonDown(interactButton);
            accumulator.sprint |= SafeButton(sprintButton);
            accumulator.crouch |= SafeButton(crouchButton);
            accumulator.pause |= SafeKeyDown(KeyCode.Escape);
        }

        private void FlushTick()
        {
            if (!EnsureTransport())
                return;

            var frameCount = Mathf.Max(1, framesAccumulated);
            var actions = new List<XaceInputAction>(12);

            AddAxis2D(actions, "move", accumulator.moveX / frameCount, accumulator.moveY / frameCount, sendIdleMovement);
            AddAxis2D(actions, "look", accumulator.lookX / frameCount, accumulator.lookY / frameCount, false);
            AddButton(actions, "jump", accumulator.jump);
            AddButton(actions, "primary_fire", accumulator.primaryFire);
            AddButton(actions, "secondary_fire", accumulator.secondaryFire);
            AddButton(actions, "interact", accumulator.interact);
            AddButton(actions, "sprint", accumulator.sprint);
            AddButton(actions, "crouch", accumulator.crouch);
            AddButton(actions, "pause", accumulator.pause);

            actions.Sort((left, right) => string.CompareOrdinal(left.action, right.action));

            var packetTick = Math.Max(latestRuntimeTick + 1, lastSentTick + 1);
            var packet = new XaceInputPacket
            {
                peer_id = peerId == 0 ? 1 : peerId,
                player_id = playerId,
                tick = packetTick,
                sequence_id = transport.NextSequenceId(),
                actions = actions.ToArray(),
                timestamp_ms = (ulong)Math.Max(0.0f, Time.realtimeSinceStartup * 1000.0f),
                device_id = string.IsNullOrWhiteSpace(deviceId) ? "unity" : deviceId.Trim(),
                predicted = predicted
            };

            if (transport.SendInputPacket(packet))
            {
                lastSentTick = packetTick;
                OnInputPacketBuilt?.Invoke(packet);
            }

            accumulator.Reset();
            framesAccumulated = 0;
        }

        private void AddAxis2D(List<XaceInputAction> actions, string name, float x, float y, bool includeIdle)
        {
            x = ClampUnit(x);
            y = ClampUnit(y);
            var active = Mathf.Abs(x) > deadzone || Mathf.Abs(y) > deadzone;
            if (!active && !includeIdle)
                return;

            actions.Add(new XaceInputAction
            {
                action = name,
                value = active ? x : 0f,
                secondary_value = active ? y : 0f,
                kind = "axis_2d",
                phase = active ? "changed" : "cancelled"
            });
        }

        private static void AddButton(List<XaceInputAction> actions, string name, bool pressed)
        {
            if (!pressed)
                return;
            actions.Add(new XaceInputAction
            {
                action = name,
                value = 1f,
                secondary_value = 0f,
                kind = "button",
                phase = "performed"
            });
        }

        private static float SafeAxis(string axisName)
        {
            if (string.IsNullOrWhiteSpace(axisName))
                return 0f;
            try { return ClampUnit(Input.GetAxisRaw(axisName)); }
            catch (Exception ex) when (IsInputUnavailable(ex)) { return 0f; }
        }

        private static bool SafeButton(string buttonName)
        {
            if (string.IsNullOrWhiteSpace(buttonName))
                return false;
            try { return Input.GetButton(buttonName); }
            catch (Exception ex) when (IsInputUnavailable(ex)) { return false; }
        }

        private static bool SafeButtonDown(string buttonName)
        {
            if (string.IsNullOrWhiteSpace(buttonName))
                return false;
            try { return Input.GetButtonDown(buttonName); }
            catch (Exception ex) when (IsInputUnavailable(ex)) { return false; }
        }

        private static bool SafeKeyDown(KeyCode key)
        {
            try { return Input.GetKeyDown(key); }
            catch (Exception ex) when (IsInputUnavailable(ex)) { return false; }
        }

        private static bool IsInputUnavailable(Exception ex)
        {
            return ex is ArgumentException || ex is InvalidOperationException;
        }

        private bool EnsureTransport()
        {
            if (transport == null)
                transport = GetComponent<XaceTransport>();
            return transport != null;
        }

        private static float ClampUnit(float value)
        {
            if (float.IsNaN(value) || float.IsInfinity(value))
                return 0f;
            return Mathf.Clamp(value, -1f, 1f);
        }

        private sealed class InputAccumulator
        {
            public float moveX;
            public float moveY;
            public float lookX;
            public float lookY;
            public bool jump;
            public bool primaryFire;
            public bool secondaryFire;
            public bool interact;
            public bool sprint;
            public bool crouch;
            public bool pause;

            public void Reset()
            {
                moveX = 0f;
                moveY = 0f;
                lookX = 0f;
                lookY = 0f;
                jump = false;
                primaryFire = false;
                secondaryFire = false;
                interact = false;
                sprint = false;
                crouch = false;
                pause = false;
            }
        }
    }
}

File: adapters/unreal/XaceInputCollector.cpp
#include "XaceInputCollector.h"

#include "Engine/World.h"
#include "GameFramework/PlayerController.h"

UXaceInputCollectorComponent::UXaceInputCollectorComponent()
{
	PrimaryComponentTick.bCanEverTick = true;
}

void UXaceInputCollectorComponent::BeginPlay()
{
	Super::BeginPlay();
	TickInterval = 1.0f / FMath::Max(1.0f, FallbackTickRateHz);
	SubscribeTransport();
}

void UXaceInputCollectorComponent::EndPlay(const EEndPlayReason::Type EndPlayReason)
{
	UnsubscribeTransport();
	Super::EndPlay(EndPlayReason);
}

void UXaceInputCollectorComponent::TickComponent(float DeltaTime, ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction)
{
	Super::TickComponent(DeltaTime, TickType, ThisTickFunction);
	if (!EnsureTransport() || !Transport->IsConnected())
	{
		return;
	}

	SampleFrame();
	FramesAccumulated++;
	TickAccumulator += DeltaTime;
	while (TickAccumulator >= TickInterval)
	{
		TickAccumulator -= TickInterval;
		FlushTick();
	}
}

void UXaceInputCollectorComponent::FlushNow()
{
	if (!EnsureTransport())
	{
		return;
	}
	SubscribeTransport();
	FlushTick();
}

void UXaceInputCollectorComponent::BindTransportNow()
{
	SubscribeTransport();
}

void UXaceInputCollectorComponent::OnRuntimeTickSnapshot(const FXaceTickSnapshot& Snapshot)
{
	LatestRuntimeTick = FMath::Max(LatestRuntimeTick, Snapshot.Tick);
}

void UXaceInputCollectorComponent::OnRuntimeHandshake(const FXaceHandshakeAck& Ack)
{
	if (Ack.TickRate > 0)
	{
		TickInterval = 1.0f / float(Ack.TickRate);
	}
}

void UXaceInputCollectorComponent::SampleFrame()
{
	APlayerController* Controller = GetWorld() != nullptr ? GetWorld()->GetFirstPlayerController() : nullptr;
	if (Controller == nullptr)
	{
		return;
	}

	MoveX += SafeAxis(Controller, MoveXAxis);
	MoveY += SafeAxis(Controller, MoveYAxis);
	LookX += SafeAxis(Controller, LookXAxis);
	LookY += SafeAxis(Controller, LookYAxis);
	bJump |= SafeKey(Controller, JumpKey);
	bPrimaryFire |= SafeKey(Controller, PrimaryFireKey);
	bSecondaryFire |= SafeKey(Controller, SecondaryFireKey);
	bInteract |= SafeKey(Controller, InteractKey);
	bSprint |= SafeKey(Controller, SprintKey);
	bCrouch |= SafeKey(Controller, CrouchKey);
	bPause |= SafeKey(Controller, EKeys::Escape);
}

void UXaceInputCollectorComponent::FlushTick()
{
	if (Transport == nullptr)
	{
		ResetAccumulated();
		return;
	}

	const float Divisor = float(FMath::Max(1, FramesAccumulated));
	TArray<FXaceInputAction> Actions;
	AddAxis2D(Actions, TEXT("move"), MoveX / Divisor, MoveY / Divisor, bSendIdleMovement);
	AddAxis2D(Actions, TEXT("look"), LookX / Divisor, LookY / Divisor, false);
	AddButton(Actions, TEXT("jump"), bJump);
	AddButton(Actions, TEXT("primary_fire"), bPrimaryFire);
	AddButton(Actions, TEXT("secondary_fire"), bSecondaryFire);
	AddButton(Actions, TEXT("interact"), bInteract);
	AddButton(Actions, TEXT("sprint"), bSprint);
	AddButton(Actions, TEXT("crouch"), bCrouch);
	AddButton(Actions, TEXT("pause"), bPause);
	Actions.Sort([](const FXaceInputAction& Left, const FXaceInputAction& Right) {
		return Left.Action < Right.Action;
	});

	FXaceInputPacket Packet;
	Packet.PeerId = FMath::Max<int64>(1, PeerId);
	Packet.PlayerId = PlayerId;
	Packet.Tick = FMath::Max(LatestRuntimeTick + 1, LastSentTick + 1);
	Packet.SequenceId = Transport->NextSequenceId();
	Packet.Actions = Actions;
	Packet.TimestampMs = int64(FPlatformTime::Seconds() * 1000.0);
	Packet.DeviceId = DeviceId.IsEmpty() ? TEXT("unreal") : DeviceId.Left(64);
	Packet.bPredicted = bPredicted;

	if (Transport->SendInputPacket(Packet))
	{
		LastSentTick = Packet.Tick;
		OnInputPacketBuilt.Broadcast(Packet);
	}
	ResetAccumulated();
}

void UXaceInputCollectorComponent::ResetAccumulated()
{
	MoveX = MoveY = LookX = LookY = 0.0f;
	bJump = bPrimaryFire = bSecondaryFire = bInteract = bSprint = bCrouch = bPause = false;
	FramesAccumulated = 0;
}

void UXaceInputCollectorComponent::AddAxis2D(TArray<FXaceInputAction>& Actions, const FString& Name, float X, float Y, bool bIncludeIdle) const
{
	X = FMath::IsFinite(X) ? FMath::Clamp(X, -1.0f, 1.0f) : 0.0f;
	Y = FMath::IsFinite(Y) ? FMath::Clamp(Y, -1.0f, 1.0f) : 0.0f;
	const bool bActive = FMath::Abs(X) > Deadzone || FMath::Abs(Y) > Deadzone;
	if (!bActive && !bIncludeIdle)
	{
		return;
	}

	FXaceInputAction Action;
	Action.Action = Name;
	Action.Value = bActive ? X : 0.0f;
	Action.SecondaryValue = bActive ? Y : 0.0f;
	Action.Kind = TEXT("axis_2d");
	Action.Phase = bActive ? TEXT("changed") : TEXT("cancelled");
	Actions.Add(Action);
}

void UXaceInputCollectorComponent::AddButton(TArray<FXaceInputAction>& Actions, const FString& Name, bool bPressed)
{
	if (!bPressed)
	{
		return;
	}
	FXaceInputAction Action;
	Action.Action = Name;
	Action.Value = 1.0f;
	Action.Kind = TEXT("button");
	Action.Phase = TEXT("performed");
	Actions.Add(Action);
}

float UXaceInputCollectorComponent::SafeAxis(APlayerController* Controller, FName AxisName) const
{
	if (Controller == nullptr || AxisName.IsNone())
	{
		return 0.0f;
	}
	return FMath::Clamp(Controller->GetInputAxisValue(AxisName), -1.0f, 1.0f);
}

bool UXaceInputCollectorComponent::SafeKey(APlayerController* Controller, const FKey& Key) const
{
	return Controller != nullptr && Key.IsValid() && Controller->IsInputKeyDown(Key);
}

bool UXaceInputCollectorComponent::EnsureTransport()
{
	if (Transport == nullptr && GetOwner() != nullptr)
	{
		Transport = GetOwner()->FindComponentByClass<UXaceTransportComponent>();
	}
	return Transport != nullptr;
}

void UXaceInputCollectorComponent::SubscribeTransport()
{
	if (bTransportSubscribed || !EnsureTransport())
	{
		return;
	}
	Transport->OnTickSnapshot.AddDynamic(this, &UXaceInputCollectorComponent::OnRuntimeTickSnapshot);
	Transport->OnHandshakeAccepted.AddDynamic(this, &UXaceInputCollectorComponent::OnRuntimeHandshake);
	bTransportSubscribed = true;
}

void UXaceInputCollectorComponent::UnsubscribeTransport()
{
	if (!bTransportSubscribed || Transport == nullptr)
	{
		return;
	}
	Transport->OnTickSnapshot.RemoveDynamic(this, &UXaceInputCollectorComponent::OnRuntimeTickSnapshot);
	Transport->OnHandshakeAccepted.RemoveDynamic(this, &UXaceInputCollectorComponent::OnRuntimeHandshake);
	bTransportSubscribed = false;
}
